File: 0386-lexicographical-numbers/0386-lexicographical-numbers.cpp
class Solution {
public:
    vector<int> lexicalOrder(int n) {
        vector<int> res;
        for (int i = 1; i <= 9 && i <= n; i++) {
            dfs(i, n, res);
        }

        return res;
    }

private:
    void dfs(int curr, int n, vector<int> &res) {
        res.push_back(curr);

        for (int i = 0; i <= 9; i++) {
            if (curr * 10 + i <= n) {
                dfs(curr * 10 + i, n, res);
            }
        }
    }
}; 

File: 0386-lexicographical-numbers/0386-lexicographical-numbers.java
class Solution {
    public List<Integer> lexicalOrder(int n) {
        List<Integer> ans = new ArrayList<>();

        for (int i = 1; i <= 9; i++) {
            generateNumbers(Integer.toString(i), n, ans);
        }
        
        return ans;
    }

    private void generateNumbers(String curr, int maxNum, List<Integer> ans) {
        int currNum = Integer.parseInt(curr);

        if (currNum > maxNum || currNum == 0) {
            return;
        }

        if (currNum != 0) {
            ans.add(currNum);
        }

        for (int i = 0; i <= 9; i++) {
            generateNumbers(curr + i, maxNum, ans);
        }
    }
}